lib/iris/experimental/ugrid.py

```diff
@@ -2894,43 +2899,122 @@ def __eq__(self, other):
         Mesh, according to the location and axis.
 
         Returns:
-        * points, bounds (array or None)
+        * points, bounds (array or None):
             lazy arrays which calculate the correct points and bounds from the
             Mesh data, based on the location and axis.
             The Mesh coordinates accessed are not identified on construction,
             but discovered from the Mesh at the time of calculation, so that
             the result is always based on current content in the Mesh.
 
         """
-        # TODO: cheat for now : correct calculations, but not dynamic.
-        # TODO: replace wth fully dynamic lazy calcs (slightly more complex).
         mesh, location, axis = self.mesh, self.location, self.axis
-        # N.B. mesh.coord returns a dict
-        node_coords = self.mesh.coord(include_nodes=True, axis=axis)
-        (node_coord,) = list(node_coords.values())
+        node_coord = _one_meshcoord(mesh.coord(include_nodes=True, axis=axis))
+
+        points_coord = None
+        bounds_connectivity = None
         if location == "node":
-            points = node_coord.core_points()
-            bounds = None
+            points_coord = node_coord
         elif location == "edge":
-            # N.B. mesh.coord returns a dict
-            edge_coords = self.mesh.coord(include_edges=True, axis=self.axis)
-            (edge_coord,) = list(edge_coords.values())
-            points = edge_coord.core_points()
-            inds = mesh.edge_node_connectivity.core_indices()
-            bounds = node_coord.core_points()[inds]
+            points_coord = _one_meshcoord(
+                mesh.coord(include_edges=True, axis=axis)
+            )
+            bounds_connectivity = mesh.edge_node_connectivity
         elif location == "face":
-            # N.B. mesh.coord returns a dict
-            face_coords = self.mesh.coord(include_faces=True, axis=self.axis)
-            (face_coord,) = list(face_coords.values())
-            points = face_coord.core_points()  # For now, this always exists
-            inds = mesh.face_node_connectivity.core_indices()
-            bounds = node_coord.core_points()[inds]
+            points_coord = _one_meshcoord(
+                mesh.coord(include_faces=True, axis=axis)
+            )
+            bounds_connectivity = mesh.face_node_connectivity
+
+        # The points output is the points of the relevant element-type coord.
+        points = points_coord.core_points()
+        if bounds_connectivity is None:
+            bounds = None
+        else:
+            # Bounds are calculated from a connectivity and the node points.
+            # Data can be real or lazy, so operations must work in Dask, too.
+            indices = bounds_connectivity.core_indices()
+            node_points = node_coord.core_points()
+            lazy = hasattr(indices, "compute") or hasattr(
+                node_points, "compute"
+            )
```

Comment:
@pp-mo `iris._lazy_data.is_lazy_data` is a handy convenience function... if you were to buy into that, then this would give:
```python
lazy = is_lazy_data(indices) or is_lazy_data(node_points)
```
which reads quite nicely

Reply:
:+1:

Reply:
done in #4053 (now here)